From 265add370068ad12068b44393763df2afd7c57ff Mon Sep 17 00:00:00 2001
From: Rene Kita <mail@rkta.de>
Date: Fri, 17 Nov 2023 16:42:48 +0100
Subject: Check for buffer underruns in checkType

This fixes GitHub Issue #282. Some buffer underrun found by some fuzzer.
---
 etc.c |   46 ++++++++++++++++++++++++++++++++--------------
 1 file changed, 32 insertions(+), 14 deletions(-)

--- a/etc.c
+++ b/etc.c
@@ -388,15 +388,23 @@ checkType(Str s, Lineprop **oprop, Linec
 			}
 			else {
 			    Strshrink(s, plen);
-			    prop -= plen;
+			    if (s->length) {
+				prop -= plen;
 #ifdef USE_ANSI_COLOR
-			    if (color)
-				color -= plen;
+				if (color)
+				    color -= plen;
 #endif
-			    if (plens == plens_buffer)
-				plen = 0;
-			    else
 				plen = *(--plens);
+			    }
+			    else {
+				plen = 0;
+				plens = plens_buffer;
+				prop = prop_buffer;
+#ifdef USE_ANSI_COLOR
+				if (color)
+				    color = color_buffer;
+#endif
+			    }
 			    str += 2;
 			}
 		    }
@@ -417,15 +425,23 @@ checkType(Str s, Lineprop **oprop, Linec
 			}
 			else {
 			    Strshrink(s, plen);
-			    prop -= plen;
+			    if (s->length) {
+				prop -= plen;
 #ifdef USE_ANSI_COLOR
-			    if (color)
-				color -= plen;
+				if (color)
+				    color -= plen;
 #endif
-			    if (plens == plens_buffer)
-				plen = 0;
-			    else
 				plen = *(--plens);
+			    }
+			    else {
+				plen = 0;
+				plens = plens_buffer;
+				prop = prop_buffer;
+#ifdef USE_ANSI_COLOR
+				if (color)
+				    color = color_buffer;
+#endif
+			    }
 			    str++;
 			}
 #else
@@ -435,10 +451,12 @@ checkType(Str s, Lineprop **oprop, Linec
 			}
 			else {
 			    Strshrink(s, 1);
-			    prop--;
+			    prop = prop == prop_buffer ? prop_buffer : prop - 1;
 #ifdef USE_ANSI_COLOR
 			    if (color)
-				color--;
+				color = (color == color_buffer
+					 ? color_buffer
+					 : color - 1);
 #endif
 			    str++;
 			}
